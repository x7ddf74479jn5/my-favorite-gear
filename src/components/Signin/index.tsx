import Container from "@material-ui/core/Container";
import { makeStyles } from "@material-ui/core/styles";
import Typography from "@material-ui/core/Typography";
import { FirebaseContext, UserContext } from "contexts";
import firebase from "firebase/app";
import paths from "paths";
import type { FC } from "react";
import React, { useContext } from "react";
import StyledFirebaseAuth from "react-firebaseui/StyledFirebaseAuth";
import { useHistory } from "react-router";

const useStyles = makeStyles((theme) => {
  return {
    root: {
      marginTop: theme.spacing(1),
      marginBottom: theme.spacing(1),
      padding: theme.spacing(2),
      backgroundColor: theme.palette.background.paper,
    },
  };
});

const Signin: FC = () => {
  const classes = useStyles();

  const { auth } = useContext(FirebaseContext);
  const { setCredential } = useContext(UserContext);
  const history = useHistory();
  const uiConfig: firebaseui.auth.Config = {
    signInFlow: "redirect",
    signInOptions: [
      {
        provider: firebase.auth.TwitterAuthProvider.PROVIDER_ID,
        customParameters: { lang: "ja" },
      },
    ],
    callbacks: {
      signInSuccessWithAuthResult: (authResult, redirectUrl) => {
        setCredential(authResult as firebase.auth.UserCredential);
        const dest = redirectUrl || paths.home;
        history.replace(dest);

        return false;
      },
    },
  };

  return (
    <Container maxWidth="xs" className={classes.root}>
      <Typography align="center" gutterBottom variant="h4">
        ログイン
      </Typography>
      <Typography paragraph align="center" color="textSecondary">
        「My Favorite
        Gear」のご利用にはログインが必須です。Twitterアカウントでログインできます。
      </Typography>
      <StyledFirebaseAuth uiConfig={uiConfig} firebaseAuth={auth} />
    </Container>
  );
};

export default Signin;
