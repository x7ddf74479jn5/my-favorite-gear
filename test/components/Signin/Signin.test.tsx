import Signin from "components/Signin";
import React from "react";

import { render } from "../../test-utils";

jest.mock("react-firebaseui/StyledFirebaseAuth", () => {
  const StyledFirebaseAuth = () => {
    return <div>StyledFirebaseAuth</div>;
  };
  return StyledFirebaseAuth;
});

jest.mock("firebase/app", () => {
  const firebase = {
    auth: {
      TwitterAuthProvider: {
        PROVIDER_ID: "",
      },
    },
  };
  return firebase;
});

describe("Signin", () => {
  it("matches snapshot", () => {
    const renderResult = render(<Signin />);
    expect(renderResult.asFragment()).toMatchSnapshot();
  });

  it("renders correctly", () => {
    const renderResult = render(<Signin />);

    expect(renderResult.container.firstElementChild).toHaveAttribute(
      "class",
      expect.stringContaining("root")
    );
    expect(renderResult.getByRole("heading", { level: 4 }));
    expect(renderResult.container.getElementsByTagName("p")).toHaveLength(1);
  });
});
