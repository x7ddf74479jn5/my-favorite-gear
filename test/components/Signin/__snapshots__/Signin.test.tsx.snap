// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signin matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root makeStyles-root-1 MuiContainer-maxWidthXs"
  >
    <h4
      class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom MuiTypography-alignCenter"
    >
      ログイン
    </h4>
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-paragraph MuiTypography-alignCenter"
    >
      「My Favorite Gear」のご利用にはログインが必須です。Twitterアカウントでログインできます。
    </p>
    <div>
      StyledFirebaseAuth
    </div>
  </div>
</DocumentFragment>
`;
